gazelle_haskell_modules/lang.go

```diff
@@ -208,37 +208,45 @@ func (*gazelleHaskellModulesLang) Fix(c *config.Config, f *rule.File) {
 		ruleNameSet[rName] = true
 	}
 
+	deleted := make(map[string]bool)
+
 	for _, r := range f.Rules {
 		if !r.ShouldKeep() && r.Kind() == "haskell_module" {
 			if _, ok := ruleNameSet[r.Name()]; !ok {
 				r.Delete()
+				deleted[r.Name()] = true
 			}
 		}
+	}
+
+	for _, r := range f.Rules {
 		if shouldModularize(r) && isNonHaskellModule(r.Kind()) {
-			cleanupModulesList(r, ruleNameSet)
-			cleanupHiddenModulesList(r, ruleNameSet)
+			cleanupModulesList(r, deleted)
+			cleanupHiddenModulesList(r, deleted)
 		}
 	}
 
 	f.Sync()
 }
 
-func cleanupModulesList(r *rule.Rule, ruleNameSet map[string]bool) {
+func cleanupModulesList(r *rule.Rule, deleted map[string]bool) {
 	// TODO: use labels instead of manually stripping away the ':' ?
 	shouldKeep := func(module string) bool {
 		if len(module) > 0 && module[0] == byte(':') {
-			return ruleNameSet[module[1:]]
+			return !deleted[module[1:]]
 		} else {
 			return true
 		}
 	}
 	cleanupModulesLists(r, "modules", shouldKeep)
 }
 
-func cleanupHiddenModulesList(r *rule.Rule, ruleNameSet map[string]bool) {
+func cleanupHiddenModulesList(r *rule.Rule, deleted map[string]bool) {
 	ruleName := r.Name()
 	// TODO: use something better?
-	shouldKeep := func(module string) bool { return ruleNameSet[fmt.Sprintf("%s.%s", ruleName, module)] }
+	shouldKeep := func(module string) bool {
+		return !deleted[fmt.Sprintf("%s.%s", ruleName, module)]
```

Comment:
> is absent from r.AttrStrings("deps")

... or rather is absent from `r.AttrStrings("deps")` before `cleanupModulesList` is called on `r`.

Reply:
I don't think I follow

0. When we're deleting `hidden_module`s from a `haskell_library` the `deps` will be listing other libraries, not modules - hence that string will always be absent from `deps`.
1. We can't do the check before calling `cleanupModulesList`, since we need `module` to create the `fmt.Sprintf(":%s.%s", ruleName, module)` string.

Am I misunderstanding something?

Reply:
>  hence that string will always be absent from deps.

You are right. I should have written `modules` instead of `deps`.

> We can't do the check before calling cleanupModulesList

It looks to me like `cleanupModulesList` and `cleanupHiddenModulesList` act on different attributes and are therefore callable in any order. Nonetheless, one can still preserve the order and capture the value of `r.AttrStrings("module")` before the first call. Anyway, I gave it a go in 6a493d9. Let me know what you think.

Reply:
Ah I see. Seems ok, thanks :)